From 1c5e1138d895c0bee06a5d2786e4ea77e41f8545 Mon Sep 17 00:00:00 2001
From: Hans-Kristian Arntzen <post@arntzen-software.no>
Date: Wed, 26 Aug 2026 13:02:09 +0200
Subject: [PATCH 04/19] vkd3d: Properly migrate a command queue to out of band
 queue.

Could enter stagger or weird scheduling territory if not.

Signed-off-by: Hans-Kristian Arntzen <post@arntzen-software.no>
---
 libs/vkd3d/command_queue_vkd3d_ext.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/libs/vkd3d/command_queue_vkd3d_ext.c b/libs/vkd3d/command_queue_vkd3d_ext.c
index ac970e88..e318dc63 100644
--- a/libs/vkd3d/command_queue_vkd3d_ext.c
+++ b/libs/vkd3d/command_queue_vkd3d_ext.c
@@ -74,6 +74,23 @@ static HRESULT STDMETHODCALLTYPE d3d12_command_queue_vkd3d_ext_NotifyOutOfBandCo
                 command_queue->device->queue_families[i]->out_of_band_queue)
         {
             struct vkd3d_queue *out_of_band_queue = command_queue->device->queue_families[i]->out_of_band_queue;
+
+            /* Migrate the command queue to the out of band queue, so that it cannot conflict
+             * with other queues from scheduling PoV.
+             * The assumption is that NvAPI notifies before the queue has actually been used
+             * for anything meaningful. */
+            d3d12_device_unmap_vkd3d_queue(command_queue->vkd3d_queue, command_queue);
+            pthread_mutex_lock(&out_of_band_queue->command_queue_mutex);
+            {
+                out_of_band_queue->virtual_queue_count++;
+                vkd3d_array_reserve(
+                        (void**)&out_of_band_queue->command_queues, &out_of_band_queue->command_queue_size,
+                        out_of_band_queue->command_queue_count + 1,
+                        sizeof(*out_of_band_queue->command_queues));
+                out_of_band_queue->command_queues[out_of_band_queue->command_queue_count++] = command_queue;
+            }
+            pthread_mutex_unlock(&out_of_band_queue->command_queue_mutex);
+
             vkd3d_set_queue_out_of_band(command_queue->device, out_of_band_queue, vk_queue_type);
             command_queue->vkd3d_queue = out_of_band_queue;
             break;
